crates/sparrow-batch/src/batch.rs

```diff
@@ -231,47 +261,58 @@ impl Batch {
+                    // minimum up_to_time. If
+                    // 1) self.up_to_time < time_exclusive: the input may have more rows
+                    // between self.up_to_time and time_exclusive.
+                    // 2) self.up_to_time > time_exclusive: We requested a split at
+                    // time_exclusive, so we can't return a batch with a greater time.
+                    up_to_time: self.up_to_time.min(time_exclusive),
+                });
+            }
 
-        // If we reach this point, then we need to actually split the data.
-        debug_assert!(time_exclusive <= self.up_to_time);
-
-        let split = data.split_up_to(time_exclusive);
-        if let Some(data) = split {
-            // Use the new max_present_time as the up_to_time.
-            let up_to_time = data.time().value(data.time.len() - 1).into();
-            Some(Batch {
-                data: Some(data),
-                // We can be complete up to time_exclusive because it is less
-                // than the time this batch was complete to. We put
-                // all of the rows this batch had up to that time in the result,
-                // and only left the batches after that time.
-                up_to_time,
-            })
+            // If we reach this point, then we need to actually split the data.
+            debug_assert!(time_exclusive <= self.up_to_time);
+
+            let split = data.split_up_to(time_exclusive);
+            if let Some(data) = split {
+                // Use the new max_present_time as the up_to_time.
+                let up_to_time = data.time().value(data.time.len() - 1).into();
+                Some(Batch {
+                    data: Some(data),
+                    // We can be complete up to time_exclusive because it is less
+                    // than the time this batch was complete to. We put
+                    // all of the rows this batch had up to that time in the result,
+                    // and only left the batches after that time.
+                    up_to_time,
+                })
+            } else {
+                // If there's no data the min_present_time must have been
+                // after the split time. We can safely return None without
+                // handling any up_to_time.
+                None
+            }
         } else {
-            // If there's no data after we split at `time_exclusive`, return None
-            None
+            // No data, but we may have to produce an empty batch with
+            // a new up_to_time. Use the minimum, as
+            // 1) self.up_to_time < time_exclusive: the input may have more rows
+            // between self.up_to_time and time_exclusive.
+            // 2) self.up_to_time > time_exclusive: We requested a split at
+            // time_exclusive, so we can't return a batch with a greater time.
+            let new_up_to = time_exclusive.min(self.up_to_time);
+            Some(Batch::new_empty(new_up_to))
```

Comment:
It would be nice if we could split this logic out of `Batch`, and track watermarks separately

Reply:
Yeah. I know we've discussed the fact that we should separate `Batch` (and how data flows) from the watermark / `up_to_time`. Do we have a ticket we should reference, or should we create one?

Reply:
Added reference to issue